docs/reusing-typecheck-results.md

```diff
@@ -0,0 +1,251 @@
+--------------------------------------------------------------------------------------------------------
+```
+
+The phases taking the longest are marked in the right. Those are assembly import, type check, optimization and IL writing.
+
+## Implementation plan
+
+The conclusion from the above is that there is a lot of potential in caching - at the same time, implementing things to the full potential right away will require serious changes in the compiler code, which will be dangerous, hard to test and review. Therefore, I propose to implement the feature in stages where each stage brings gains in some scenarios.
+
+**Stage 1 - force-gen and compare typechecking graphs**
+
+We already create the TC graph in `main1` and we are able to dump it (in Mermaid) via the compiler flags.
+
+So we can do the following:
+- force-gen and save the graph
+- append all the extra compilation info (basically, the compilation arguments string)
+- before doing the retypecheck, detect if there is any change to above:
+  - if so, skip the retypecheck
+  - otherwise, apply some algorithm to get the graph diff. The result should be the list of files for retypecheck (just ignore it for now)
+
+This will allow to skip a huge part of the compilation in the scenarios like reopening Visual Studio for a project when nothing has changed.
+
+**Stage 2 - skip retypecheck for some files**
+
+In `main1`, we do unpickling (= deserializing) and pickling (= serializing) of the code. This currently happens on the module/namespace basis. 
+
+So here we can do the following:
+- extend the pickler logic so that it can work on the file basis
+- save acquired typecheck information to the intermediate folder after the typecheck
+- for the files not needing retypecheck as per graph diff, skip the phase, instead restore the typecheck info
+
+This will hugely benefit the scenarios when only the edge of the graph is affected (think one test in a test project).
+
+**Stage 3 - reduce signature data generation**
+
+In `main3`, we encode signature data in the assemblies.
+
+Since we'll have typecheck graph at this point, we can detect its "edges" - files which nothing depends on. Those don't need to include signature data, hence we can skip the encoding step for them.
+
+Here, the gain largely depends on the type of the project, but this would be anyway a good move in the direction of the compiler less relying on signatures for its operations.
+
+**Stage 4 - reuse import data**
+
+In `main1`, we restore imported assemblies to get their IL (e.g. system assemblies).
+
+So here we can:
+- serialize and cache imported IL on the per-assembly basis
+- apply this for the cross-project case, to not reimport IL for repeating assemblies but instead restore it from the cache
+
+This will be a smaller gain for any particular project but a big accummulated one for large multi-project solutions.
```

Comment:
```suggestion
This will be a smaller gain for any particular project but a big accummulated one for large multi-project solutions.

This step includes adding a cross-project intermmediate file location to be coordinated with msbuild properties.
```